# Práctica 2. Resolución de problemas con búsqueda heurística
## Ingeniería del Conocimiento    2025/2026
### Prof. Juan A. Recio García

## Parte II: Búsqueda Heurística

In [63]:
from search import *

# EJERCICIO: 8 Puzzle con búsqueda heurística

#### Para el problema del puzle de 8  vamos a definir las siguientes funciones heurísticas:
* linear(node): cuenta el número de casillas mal colocadas respecto al estado final.
* manhattan(node): suma la distancia Manhattan desde cada casilla a la posición en la que debería estar en el estado final.
* max_heuristic(node): maximo de las dos anteriores
* sqrt_manhattan(node):  raíz cuadrada de la distancia Manhattan

In [64]:
from math import sqrt
class Ocho_Puzzle(Problem):
    """Problema a del 8-puzzle.  Los estados serán tuplas de nueve elementos,
    permutaciones de los números del 0 al 8 (el 0 es el hueco). Representan la
    disposición de las fichas en el tablero, leídas por filas de arriba a
    abajo, y dentro de cada fila, de izquierda a derecha. 
    
    Las cuatro acciones del problema las representaremos mediante las cadenas:
    "Mover hueco arriba", 
    "Mover hueco abajo", 
    "Mover hueco izquierda" y
    "Mover hueco derecha"."""""

    def __init__(self, initial, goal=(1, 2, 3, 4, 5, 6, 7, 8, 0)):
        """ Define goal state and initialize a problem """
        self.goal = goal
        Problem.__init__(self, initial, goal)
        self.acc = ['Mover hueco arriba', 'Mover hueco abajo', 'Mover hueco derecha', 'Mover hueco izquierda']

    def actions(self, estado):
        return [a for a in self.acc if(self.result(estado, a) != "No valido")]

    def result(self, estado, accion):
        pos_hueco = estado.index(0)
        l = list(estado)
        def swap(x):
            l[pos_hueco], l[pos_hueco+x] = l[pos_hueco+x], l[pos_hueco]
        
        if accion == 'Mover hueco arriba':
            if pos_hueco >= 3:
                swap(-3)
            else:
                return "No valido"
        elif accion == 'Mover hueco abajo':
            if pos_hueco <= 5:
                swap(3)
            else:
                return "No valido"
        elif accion == 'Mover hueco derecha':
            if pos_hueco not in [2, 5, 8]:
                swap(1)
            else:
                return "No valido"
        elif accion == 'Mover hueco izquierda':
            if pos_hueco not in [0, 3, 6]:
                swap(-1)
            else:
                return "No valido"

        return tuple(l)

    # Hemos añadido las heurísticas a la clase para poder acceder al atributo goal
    def linear(self, node):
        return count([1 for index, val in enumerate(node.state) if val - 1 != index])

    def manhattan(self, node):
        distancia = 0
        for indice, valor in enumerate(node.state):
            if valor == 0:
                continue
            
            fila_actual = indice // 3
            col_actual = indice % 3
            
            fila_objetivo = (valor - 1) // 3
            col_objetivo = (valor - 1) % 3
            
            distancia += abs(fila_actual - fila_objetivo) + abs(col_actual - col_objetivo)
            
        return distancia

    def sqrt_manhattan(self, node):
        return sqrt(self.manhattan(node))

    def max_heuristic(self, node):
        return max(self.linear(node), self.manhattan(node))

    def h(self, node):
        """ Return the heuristic value for a given state."""
        return self.max_heuristic(node)
    
    def coste_de_aplicar_accion(self, s, a):
        return self.h(self.result(s, a))

#### Vamos a usar las implementaciones de AIMA de los algoritmos de  busqueda_primero_el_mejor y búsqueda_a_estrella (con las heurísticas anteriores) 




#### Hay que comparar los costes temporales usando %timeit y comentar los resultados.

In [65]:
####  Vamos a probar el 8-puzzle utilizando el siguiente **estado inicial** 

#              +---+---+---+
#              | 2 | 4 | 3 |
#              +---+---+---+
#              | 1 | 5 | 6 |
#              +---+---+---+
#              | 7 | 8 | H |
#              +---+---+---+

puzle = Ocho_Puzzle((2, 4, 3, 1, 5, 6, 7, 8, 0)) 

In [66]:
# Función para mostrar el estado del tablero de forma más visual
def show(s):
    print(" ".join(map(str,s[:3])))
    print(" ".join(map(str,s[3:6])))
    print(" ".join(map(str,s[6:])))

In [67]:
show(puzle.initial)

2 4 3
1 5 6
7 8 0


In [68]:
show(puzle.goal)

1 2 3
4 5 6
7 8 0


### Ahora probamos los algoritmos de búsqueda informada con la heurística Manhatan

In [69]:
# Búsqueda voraz con la heurística de Manhattan
greedy_best_first_graph_search(puzle, puzle.manhattan).solution()

['Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco abajo',
 'Mover hueco derecha',
 'Mover hueco derecha',
 'Mover hueco abajo']

In [70]:
# Búsqueda A* con la heurística de Manhattan
astar_search(puzle,puzle.manhattan).solution()

['Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco abajo',
 'Mover hueco derecha',
 'Mover hueco derecha',
 'Mover hueco abajo']

### Probamos las otras heurísticas

In [71]:
astar_search(puzle,puzle.linear).solution()

['Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco abajo',
 'Mover hueco derecha',
 'Mover hueco derecha',
 'Mover hueco abajo']

In [72]:
astar_search(puzle,puzle.max_heuristic).solution()

['Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco abajo',
 'Mover hueco derecha',
 'Mover hueco derecha',
 'Mover hueco abajo']

In [73]:
astar_search(puzle,puzle.sqrt_manhattan).solution()

['Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco arriba',
 'Mover hueco izquierda',
 'Mover hueco abajo',
 'Mover hueco derecha',
 'Mover hueco derecha',
 'Mover hueco abajo']

### ¿Has notado diferencias en los tiempos de ejecución? ¿y en los resultados?
Vamos a medirlo 

In [74]:
puzzle_1 = Ocho_Puzzle((2, 4, 3, 1, 5, 6, 7, 8, 0))
puzzle_2 = Ocho_Puzzle((1, 2, 3, 4, 5, 6, 0, 7, 8))
puzzle_3 = Ocho_Puzzle((1, 2, 3, 4, 5, 7, 8, 6, 0))

In [75]:
%%timeit
astar_search(puzzle_1, puzzle_1.linear)
astar_search(puzzle_2, puzzle_2.linear)
astar_search(puzzle_3, puzzle_3.linear)

1.65 ms ± 21.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [76]:
%%timeit
astar_search(puzzle_1, puzzle_1.manhattan)
astar_search(puzzle_2, puzzle_2.manhattan)
astar_search(puzzle_3, puzzle_3.manhattan)

923 μs ± 2.6 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [77]:
%%timeit
astar_search(puzzle_1, puzzle_1.sqrt_manhattan)
astar_search(puzzle_2, puzzle_2.sqrt_manhattan)
astar_search(puzzle_3, puzzle_3.sqrt_manhattan)

15 ms ± 296 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [78]:
%%timeit
astar_search(puzzle_1, puzzle_1.max_heuristic)
astar_search(puzzle_2, puzzle_2.max_heuristic)
astar_search(puzzle_3, puzzle_3.max_heuristic)

1.41 ms ± 9.67 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


####   Escribe aquí tus conclusiones sobre qué heurística te parece mejor para este problema. Para ello, ordena las heurísticas según sean más informadas que otras. P.e.: linear > manhatan > sqrt_manhattan > max_heuristic. Razona la respuesta



En la práctica, la mejor heurística para este problema concreto es manhattan.

Aunque teóricamente no es la heurística que menos nodos expande (la más informada), ofrece el equilibrio perfecto entre calidad de la estimación y coste computacional. Es la opción ganadora en tiempo de ejecución, resolviendo los puzzles en un promedio de 923 µs, siendo significativamente más rápida que las alternativas.

max_heuristic = manhattan > linear > sqrt_manhattan

manhattan:

Es una heurística muy informada. La distancia Manhattan calcula cuántos pasos exactos le faltan a cada pieza asumiendo que no hay obstáculos. Esto recorta el árbol de búsqueda drásticamente (expande pocos nodos). Como solo requiere sumas y restas, el cálculo es rapidísimo, logrando el mejor tiempo.

max_heuristic:

Por definición, si una pieza está mal colocada su distancia Manhattan será como mínimo 1 (y normalmente más), mientras que linear siempre dará. Por tanto, el valor de manhattan siempre será mayor o igual al valor de linear.

Esto significa que hacer max(linear, manhattan) devuelve siempre el valor de manhattan.

Por lo tanto, técnicamente es igual de informada que manhattan.

linear:

Es menos informada que manhattan. Saber solo si una pieza está en su sitio o no da muy poca información (una pieza mal colocada podría estar a 1 paso de su sitio o a 4, y linear devuelve 1 en ambos casos).

Al subestimar tanto el coste real, el algoritmo A* se ve obligado a expandir y explorar muchos más nodos buscando el camino óptimo.

sqrt_manhattan:

Al aplicarle la raíz cuadrada al valor de manhattan, estás aplastando el número. Si a una pieza le faltan 9 pasos, la heurística dirá que le faltan 3.

Esto la convierte en la menos informada de todas.